### PR TITLE
19936 - EFT - PAY-API changes to handle CFS when switching payment methods

### PR DESCRIPTION
**Pull request by Jxio**
*Issue #:*
https://github.com/bcgov/entity/issues/19936
*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-pay license (Apache 2.0).

### CODE REVIEW

**Comment by ochiu** [comment]
https://github.com/bcgov/sbc-pay/blob/3552acd9be9e371a77522597c9282d39be79d44e/pay-api/src/pay_api/services/payment_account.py#L481

I think the logic for _handle_payment_details() may need to be reviewed. There is some default logic for if a cfs account already exists it will do updates treating it like PAD.

`cls._update_pad_activation_date(cfs_account, is_sandbox, payment_account)`

File: pay-api/src/pay_api/services/eft_service.py [hunk 1]
```diff
@@ -35,9 +36,13 @@ class EftService(DepositService):
     """Service to manage electronic fund transfers."""
 
     def get_payment_method_code(self):
-        """Return EFT as the system code."""
+        """Return EFT as the payment method code."""
         return PaymentMethod.EFT.value
 
+    def get_payment_system_code(self):
+        """Return PAYBC as the system code."""
+        return PaymentSystem.PAYBC.value
```
Comment (seeker25) [inline pay-api/src/pay_api/services/eft_service.py:44]: EFT doesn't use PAYBC?

Comment (seeker25) [reply pay-api/src/pay_api/services/eft_service.py:44]: Would recommend putting in another enum, just call it AR?

File: pay-api/src/pay_api/services/eft_service.py [hunk 2]
```diff
@@ -35,9 +36,13 @@ class EftService(DepositService):
     """Service to manage electronic fund transfers."""
 
     def get_payment_method_code(self):
-        """Return EFT as the system code."""
+        """Return EFT as the payment method code."""
         return PaymentMethod.EFT.value
 
+    def get_payment_system_code(self):
+        """Return PAYBC as the system code."""
```
Comment (seeker25) [inline pay-api/src/pay_api/services/eft_service.py:43]: Is the AR module considered PAYBC? I thought PAYBC was just credit cards, I could be wrong

File: pay-api/tests/unit/api/test_account.py [hunk 3]
```diff
@@ -475,6 +475,39 @@ def test_update_pad_account_when_cfs_up(session, client, jwt, app):
...
 
+def test_switch_eft_account_when_cfs_down(session, client, jwt, app, admin_users_mock):
+    """Assert that the payment records are created with 202."""
+    token = jwt.create_jwt(get_claims(role=Role.SYSTEM.value), token_header)
+    headers = {'Authorization': f'Bearer {token}', 'content-type': 'application/json'}
+    rv = client.post('/api/v1/accounts',
+                     data=json.dumps(get_basic_account_payload(payment_method=PaymentMethod.PAD.value)),
+                     headers=headers)
+    auth_account_id = rv.json.get('accountId')
+    # Mock ServiceUnavailableException
```
Comment (seeker25) [inline pay-api/tests/unit/api/test_account.py:486]: redundant - could remove or include more details that the code below doesn't include

**Review by seeker25 — APPROVED** [review]

**Comment by seeker25** [comment]
Ody's comments need to be addressed before merging

File: pay-api/src/pay_api/services/eft_service.py (hunk 2 above)
Comment (Jxio) [reply pay-api/src/pay_api/services/eft_service.py:43]: I might wrong, I looked at pay_api code, the direct_pay, online banking, pad return the paybc system code, so I thought it same for eft.

Comment (seeker25) [reply pay-api/src/pay_api/services/eft_service.py:43]: no worries, sounds good

File: pay-api/src/pay_api/services/eft_service.py [hunk 4]
```diff
@@ -47,6 +52,17 @@ def create_account(self, identifier: str, contact_info: Dict[str, Any], payment_
...
         return cfs_account
 
+    def update_account(self, name: str, cfs_account: CfsAccountModel, payment_info: Dict[str, Any]) -> CfsAccountModel:
+        """Update pad account."""
+        if str(payment_info.get('bankInstitutionNumber')) != cfs_account.bank_number or \
+                str(payment_info.get('bankTransitNumber')) != cfs_account.bank_branch_number or \
+                str(payment_info.get('bankAccountNumber')) != cfs_account.bank_account_number:
+            # This means, the current cfs_account is for PAD, not EFT
+            # Make the current CFS Account as INACTIVE in DB
+            cfs_account.status = CfsAccountStatus.INACTIVE.value
```
Comment (Jxio) [inline pay-api/src/pay_api/services/eft_service.py:62]: Inactive current PAD cfs_account

File: pay-api/src/pay_api/services/payment_account.py [hunk 5]
```diff
@@ -484,7 +484,10 @@ def _handle_payment_details(cls, account_request, is_sandbox, pay_system, paymen
         cfs_account: CfsAccountModel = CfsAccountModel.find_effective_by_account_id(payment_account.id) \
             if payment_account.id else None
         if pay_system.get_payment_system_code() == PaymentSystem.PAYBC.value:
-            if cfs_account is None:
+            if cfs_account is None or (payment_account.payment_method == PaymentMethod.EFT and cfs_account):
```
Comment (Jxio) [inline pay-api/src/pay_api/services/payment_account.py:487]: when Switching from PAD to EFT, we always create a new cfs account, cause the previous EFT cfs account suppose to be inactivated when it switch to PAD, and also I can't just remove the payment details use cfs service, it returns 404.

File: pay-api/src/pay_api/services/eft_service.py [hunk 6]
```diff
@@ -47,6 +52,17 @@ def create_account(self, identifier: str, contact_info: Dict[str, Any], payment_
         cfs_account.status = CfsAccountStatus.PENDING.value
         return cfs_account
 
+    def update_account(self, name: str, cfs_account: CfsAccountModel, payment_info: Dict[str, Any]) -> CfsAccountModel:
+        """Update pad account."""
+        if str(payment_info.get('bankInstitutionNumber')) != cfs_account.bank_number or \
+                str(payment_info.get('bankTransitNumber')) != cfs_account.bank_branch_number or \
+                str(payment_info.get('bankAccountNumber')) != cfs_account.bank_account_number:
+            # This means, the current cfs_account is for PAD, not EFT
+            # Make the current CFS Account as INACTIVE in DB
```
Comment (seeker25) [inline pay-api/src/pay_api/services/eft_service.py:61]: redundant, read the two lines below

File: pay-api/src/pay_api/services/eft_service.py [hunk 7]
```diff
@@ -47,6 +52,17 @@ def create_account(self, identifier: str, contact_info: Dict[str, Any], payment_
         cfs_account.status = CfsAccountStatus.PENDING.value
         return cfs_account
 
+    def update_account(self, name: str, cfs_account: CfsAccountModel, payment_info: Dict[str, Any]) -> CfsAccountModel:
+        """Update pad account."""
+        if str(payment_info.get('bankInstitutionNumber')) != cfs_account.bank_number or \
+                str(payment_info.get('bankTransitNumber')) != cfs_account.bank_branch_number or \
+                str(payment_info.get('bankAccountNumber')) != cfs_account.bank_account_number:
+            # This means, the current cfs_account is for PAD, not EFT
```
Comment (seeker25) [inline pay-api/src/pay_api/services/eft_service.py:60]: this is good, intent of why you wrote the code

File: pay-api/src/pay_api/services/payment_account.py (hunk 5 above)
Comment (seeker25) [reply pay-api/src/pay_api/services/payment_account.py:487]: Wouldn't the above technically be something good for a comment?

File: pay-api/src/pay_api/services/eft_service.py (hunk 4 above)
Comment (seeker25) [reply pay-api/src/pay_api/services/eft_service.py:62]: Above is better than "# Make the current CFS Account as INACTIVE in DB"

**Pull request merged**